### DISCUSSION TITLE
Add support for the HEIF image format

### DISCUSSION THREAD

**Discussion by fleaz** [craftcms/cms#9115] [Ideas] [answered]
### Description

Hey,
we are running a Craft site for a customer who wants to upload images directly from his iPhone what confronted us with the problem of handling `*.heic` files (Modern iOS devices shoot images in [HEIF](https://en.wikipedia.org/wiki/High_Efficiency_Image_File_Format) instead of JPEG).

We manually compiled a recent version of imagemagick with libheif on the server to get support for this image format and after adding `heic` to `extraAllowedFileExtensions` in the general config, one can successfully upload images in the HEIF format.

But Craft won't generate a thumbnail because it probably does not know that this is an image format (there is e.g. also no height/width information).

Are there any plans on adding HEIF support to Craft? Most distributions currently don't ship a recent enough version of Imagemagick (so you have compile it yourself), but this will hopefully change in the future :D

Best,
Felix

**Comment by codyjames** [comment]
Yup, we're running into this as well. Can't really launch without this support due to the nature of our project (hundreds of images uploaded by a community each day — many of which likely coming from iphones).

**Comment by andris-sevcenko** [comment]
I did look into this and the whole effort stopped in the Imagick PHP extension not being able to determine the colorspace of the HEIF image. Interestingly, ImageMagick via CLI was able to determine that.

Curios, @fleaz, what's the output you get if you do `print_r((new Imagick('path/to/file.heif'))->getColorspace());`?

**Comment by mattgrayisok** [comment]
Also tracking this for Servd.

@andris-sevcenko Using these two images:

https://filesamples.com/samples/image/heif/sample1.heif
https://filesamples.com/samples/image/heic/sample1.heic

On PHP 8. 
ImageMagick 7.0.10-57. 
Imagick commit '448c1cd0d58ba2838b9b6dff71c9b7e70a401b90' (built from master)

The output is `0` for both images.

On CLI:

```
bash-5.1# magick identify -verbose test.heic | grep space
  Colorspace: YCbCr

bash-5.1# magick identify -verbose test.heif
Killed
```

So it looks like Image Magick CLI is working ok for identification but has some memory issues when working with moderately large (2MB in this case) heif files - it allocated ~500MB RAM before it was killed.

**Comment by andris-sevcenko** [comment]
@mattgrayisok Thanks for the additional info!

Without knowing the colorspace, any image manipulation and preserving original colors is pretty much DOA, so currently no easy way on how Craft can implement this.

One way of proceeding about this, in general, is to refactor how images are manipulated in general by Craft to allow third parties to provide image drivers, in which case it would become possible for a plugin to provide an interface with the `libheif` tool, perhaps. This, of course, would be a rather significant undertaking but one we are discussing internally.

**Comment by fleaz** [comment]
Hey @andris-sevcenko 
thanks for looking into this!

Just wanted to confirm that I get the same results as Matt on my machine (PHP 7.4, Imagick 7.0.11). Coincidentally we even used the same example images to test this :D

I also see huge memory allocation on the second image (~650-700MB) but after a while it return with `YCbCr`. The `getColorSpace()` also returns `0` for both images.

**Comment by brandonkelly** [comment, marked as answer]
Circled back on this today, and noticed the Imagine lib got updated recently to support HEIC. So I’ve pulled that in for the next Craft 3 and 4 releases. (d7ee2e20b291b055b3ef8a93cf31da1800ba6103)

**Reply by brandonkelly** [reply]
Craft 3.7.63 and 4.3.6 are out with HEIC support!

**Comment by fleaz** [comment]
Awesome. Thanks for still thinking about this \o/

**Comment by sunscreem** [comment]
@brandonkelly Is it possible to disable this in the config? We're finding clients are uploading HEIF images but they aren't supported in Chrome or our version of imagemagik.

**Reply by RyanRoberts** [reply]
> HEIF images but they aren't supported in Chrome

Or any browser? https://caniuse.com/?search=heif

**Reply by brandonkelly** [reply]
We don’t list it as a web-safe format here:

https://github.com/craftcms/cms/blob/c3afbef706beafbe71b64735e1124ff345dc6e48/src/helpers/Image.php#L150-L153

So as long as you are outputting images with a transform (which you should always do, to be safe), HEIC/HEIF images will get converted to JPEGs.

**Comment by MatthiasBrodelet** [comment]
Is anyone still having issues with this?
I keep getting "The file “IMG_28346808df6785c8d6.91828556.HEIC” does not appear to be an image." when I try to upload an image from my iphone. I a running Craft 5.7.2 with Imagick 3.7.0 (ImageMagick 6.9.11-60).

**Reply by brandonkelly** [reply]
Go to Utilities → PHP Info and search for `HEIC`. Is it listed within the “ImageMagick supported formats” output?

**Reply by MatthiasBrodelet** [reply]
Yes, it is.
Seems like this is not a Craft CMS issue but a libheif issue that is used by Imagick. Since iOS 18 Apple changed something in the HEIC files which causes these problems.

It should be fixed with libheif 1.18+.

https://github.com/Imagick/imagick/issues/675

**Reply by brandonkelly** [reply]
Ah cool, thanks for looking into it!
